### PR TITLE
Implement IfStatement, BlockStatement

### PR DESCRIPTION
**Pull request by boohyunsik**
resolved: #[108]

WIP : Implement IfStatement, BlockStatement

- [x] Test case

### CODE REVIEW

File: parse/parser_internal_test.go [hunk 1]
```diff
@@ -611,3 +611,45 @@ func TestParseInfixExpression(t *testing.T) {
...
+			{Type: Lparen, Val: "("},
+			{Type: Bool, Val: "true"},
+			{Type: Rparen, Val: ")"},
+			{Type: Lbrace, Val: "{"},
+			{Type: IntType, Val: "int"},
+			{Type: Ident, Val: "a"},
+			{Type: Rbrace, Val: "}"},
+		},
+	}
+
```
Comment (junbeomlee) [inline parse/parser_internal_test.go:1207]: i think we should test `else` part also!

Comment (boohyunsik) [reply parse/parser_internal_test.go:1207]: Yeap. This part was paused. I will do it.

File: ast/ast.go [hunk 2]
```diff
@@ -113,6 +113,34 @@ func (as *AssignStatement) String() string {
...
+type IfStatement struct {
+	Condition   Expression
+	Consequence *BlockStatement
+	Alternative *BlockStatement
+}
+
+func (is *IfStatement) do() {}
+
+func (is *IfStatement) String() string {
+	return fmt.Sprintf("if ( %s ) { %s }", is.Condition.String(), is.Consequence.String())
```
Comment (zeroFruit) [inline ast/ast.go]: we need to print else part if Alternative exist

File: parse/parser.go [hunk 3]
```diff
@@ -337,3 +337,67 @@ func parseAssignStatement(buf TokenBuffer) (*ast.AssignStatement, []error) {
 func parseCallExpression(buf TokenBuffer, fn ast.Expression) (ast.Expression, []error) {
 	return nil, nil
 }
+
+func parseIfStatement(buf TokenBuffer) (*ast.IfStatement, []error) {
+	errs := make([]error, 0)
+	var err []error
+	tok := buf.Read()
+	if tok.Type != If {
+		errs = append(errs, errors.New("parseIfStatement() error - No IF statement"))
```
Comment (zeroFruit) [inline parse/parser.go]: How about using parseError instead of just errors.New?

Comment (boohyunsik) [reply parse/parser.go]: This will be fixed.

File: ast/ast.go (hunk 2 above)
Comment (boohyunsik) [reply ast/ast.go]: Ok.

**Comment by boohyunsik** [comment]
Plz review it! Parsing is almost done :-D

File: parse/parser.go [hunk 4]
```diff
@@ -178,8 +178,18 @@ func Parse(buf TokenBuffer) (*ast.Program, error) {
 // TODO: implement me w/ test cases :-)
 func parseStatement(buf TokenBuffer) (ast.Statement, error) {
 	switch buf.Peek(CURRENT).Type {
+	case Assign:
```
Comment (zeroFruit) [inline parse/parser.go]: I think Assign should be dealt inside the parseAssignStatement, so I think this is not needed

File: parse/parser.go [hunk 5]
```diff
@@ -178,8 +178,18 @@ func Parse(buf TokenBuffer) (*ast.Program, error) {
...
+	case BoolType:
+		return parseAssignStatement(buf)
+	case If:
+		return parseIfStatement(buf)
+	case IntType:
+		return parseAssignStatement(buf)
 	case Return:
 		return parseReturnStatement(buf)
+	case StringType:
+		return parseAssignStatement(buf)
```
Comment (zeroFruit) [inline parse/parser.go]: Why don't we sort cases IntType, BoolType, StringType first. I mean sort Type cases first

File: parse/parser.go [hunk 6]
```diff
@@ -454,3 +464,83 @@ func isDataStructure(tok Token) bool {
 		tok.Type == IntType ||
 		tok.Type == BoolType
 }
+
+func parseIfStatement(buf TokenBuffer) (*ast.IfStatement, error) {
+	tok := buf.Read()
+	if tok.Type != If {
+		err := parseError{
+			tok.Type,
+			fmt.Sprintf("is not a If"),
```
Comment (zeroFruit) [inline parse/parser.go]: we can just use "" string, if we don't use variable in the string

File: parse/parser.go (hunk 5 above)
Comment (boohyunsik) [reply parse/parser.go]: Oh, I sorted alphabetically.
I will modify it.

File: parse/parser.go (hunk 4 above)
Comment (boohyunsik) [reply parse/parser.go]: But without this, there was an error.
 I will retry.

**Comment by zeroFruit** [comment]
LGTM

**Review by zeroFruit — APPROVED** [review]

**Pull request merged**